buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java

```diff
@@ -0,0 +1,141 @@
+
+            @Override
+            public void settingsEvaluated(Settings settings) {}
+
+            @Override
+            public void projectsLoaded(Gradle gradle) {}
+
+            @Override
+            public void projectsEvaluated(Gradle gradle) {}
+
+            @Override
+            public void buildFinished(BuildResult result) {
+                if (reaperProcess != null) {
+                    try {
+                        reaperProcess.getOutputStream().close();
+                        project.getLogger().info("Waiting for reaper to exit normally");
+                        if (reaperProcess.waitFor() != 0) {
+                            throw new GradleException("Reaper process failed. Check log at "
+                                + baseDirectory.resolve("error.log") + " for details");
+                        }
+                    } catch (Exception e) {
+                        throw new RuntimeException(e);
+                    }
+
+                }
+            }
+        });
+    }
+
+    /**
+     * Register a pid that will be killed by the reaper.
+     */
+    public void registerPid(String serviceId, long pid) {
+        String[] killPidCommand = OS.<String[]>conditional()
+            .onWindows(() -> new String[]{"Taskill", "/F", "/PID", String.valueOf(pid)})
+            .onUnix(() -> new String[]{"kill", "-9", String.valueOf(pid)}).supply();
+        registerCommand(serviceId, killPidCommand);
+    }
+
+    /**
+     * Register a system command that will be run by the reaper.
+     */
+    public void registerCommand(String serviceId, String... command) {
+        ensureReaperStarted();
+
+        try {
+            Files.writeString(baseDirectory.resolve(serviceId), String.join(" ", command));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public void unregister(String serviceId) {
+        try {
+            Files.deleteIfExists(baseDirectory.resolve(serviceId));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private synchronized void ensureReaperStarted() {
+        if (reaperProcess == null) {
+            try {
+                // copy the reaper jar
+                Path jarPath = project.getBuildDir().toPath().resolve("reaper").resolve("reaper.jar");
+                Files.createDirectories(jarPath.getParent());
+                InputStream jarInput = ReaperPlugin.class.getResourceAsStream("/META-INF/reaper.jar");
+                try (OutputStream out = Files.newOutputStream(jarPath)) {
+                    jarInput.transferTo(out);
+                }
+
+                // ensure the input directory exists
+                Files.createDirectories(baseDirectory);
+
+                // start the reaper
+                ProcessBuilder builder = new ProcessBuilder(
+                    "java", "-cp", jarPath.toString(), "org.elasticsearch.gradle.reaper.Reaper", baseDirectory.toString());
```

Comment:
We should also use some pretty conservative JVM memory settings here instead of going with platform defaults. We know this thing is lightweight so no need to give 1G of max heap or whatever the default is.